server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationRelocationIT.java

```diff
@@ -494,4 +495,63 @@ public void testAddNewReplicaFailure() throws Exception {
         assertTrue(clusterHealthResponse.isTimedOut());
         ensureYellow(INDEX_NAME);
     }
+
+    public void testFlushAfterRelocation() throws Exception {
+        // Starting two nodes with primary and replica shards respectively.
+        final String primaryNode = internalCluster().startNode();
+        prepareCreate(
+            INDEX_NAME,
+            Settings.builder()
+                // we want to control refreshes
+                .put("index.refresh_interval", -1)
+        ).get();
+        ensureYellowAndNoInitializingShards(INDEX_NAME);
+        final String replicaNode = internalCluster().startNode();
+        ensureGreen(INDEX_NAME);
+
+        // Start another empty node for relocation
+        final String newPrimary = internalCluster().startNode();
+        ClusterHealthResponse clusterHealthResponse = client().admin()
+            .cluster()
+            .prepareHealth()
+            .setWaitForEvents(Priority.LANGUID)
+            .setWaitForNodes("3")
+            .execute()
+            .actionGet();
+        assertEquals(clusterHealthResponse.isTimedOut(), false);
+        ensureGreen(INDEX_NAME);
+
+        // Start indexing docs and refresh index
+        final int initialDocCount = scaledRandomIntBetween(2000, 3000);
+        final List<ActionFuture<IndexResponse>> pendingIndexResponses = new ArrayList<>();
+
+        for (int i = 0; i < initialDocCount; i++) {
+            pendingIndexResponses.add(
+                client().prepareIndex(INDEX_NAME).setId(Integer.toString(i)).setSource("field", "value" + i).execute()
+            );
+        }
+        refresh(INDEX_NAME);
+
+        // Relocate primary to new primary. When new primary starts it does perform a flush.
+        logger.info("--> relocate the shard from primary to newPrimary");
+        ActionFuture<ClusterRerouteResponse> relocationListener = client().admin()
+            .cluster()
+            .prepareReroute()
+            .add(new MoveAllocationCommand(INDEX_NAME, 0, primaryNode, newPrimary))
+            .execute();
+        clusterHealthResponse = client().admin()
+            .cluster()
+            .prepareHealth()
+            .setWaitForEvents(Priority.LANGUID)
+            .setWaitForNoRelocatingShards(true)
+            .setTimeout(ACCEPTABLE_RELOCATION_TIME)
+            .execute()
+            .actionGet();
+        assertEquals(clusterHealthResponse.isTimedOut(), false);
+
+        // Verify if all docs are present in replica after flush, if new relocated primary doesn't flush after relocation the below assert
+        // will fail
```

Comment:
Basically, we need not perform any refresh and turn all aync calls to sync here. 
1. Index docs in sync without refresh so that only primary contains all the data before relocation. 
2. Assert doc count on primary is matching ingested doc count and replica should have 0 doc count
3. Relocate primary to new primary node and wait for it to complete. 
4. Assert replica gets all the docs (this should only be possible from flush operation before state switch to `STARTED`.

Reply:
Sure, makes sense. This way it will fail more definitively without flush. Thanks @dreamer-89 , I will update the test

Reply:
Asserting doc count on primary before relocation is not possible as without a refresh even on primary shard, it will have 0 docs, as we open searcher only after a refresh. So instead I am asserting on segrep stats that no replication event has taken place on replica shard before relocation